### PR TITLE
Added '%t' and '%r' tuple formats + improved Doc

### PR DESCRIPTION
**Pull request by Bolpat**
I hated using `format("(%( content %))", tup)` or `format("( content )", tup.expand)`. Something that easy should be supported by the Tuple by default.

In a nutshell:
- `%t` formats `tuple(1, "a")` like `(1, a)`; simply parenthesis around contents.
- `%r` does the same, but nested tuples are also formatted `%r`, i.e. `%r` is `%t` recursively for all subtuples. `tuple(tuple(0, 1), 2)` like `((0, 1), 2)`.

Edit: Removed Quotes around string.

### CODE REVIEW

**Comment by JakobOvrum** [comment]
AFAICS, `%t` is a shorthand for `(%(%s%|, %))`, except string elements are unquoted (maybe a bug). Format specifier real estate is precious; is `%t` really worth it?

**Comment by Bolpat** [comment]
Does the edit mean, it is ok then? That strings are not quoted is not a bug.

File: std/typecons.d [hunk 1]
```diff
@@ -791,26 +791,47 @@ template Tuple(Specs...)
...
+             * $(TROW %(P `%r`), $(P Format like `%t`. Nested tuples are formatted like `%t` recursively.))
+             * $(TROW $(P `%(`inner`%)`), $(P The format `inner` is applied the expanded Tuple, so it may
+             *      contain as many formats as the Tuple has fields.))
+             * $(TROW $(P `%(`inner`%|`sep`%)`), $(P The inner format is applied on all fields of the Tuple
+             *      and must be compatible to all of them.)))
+             * For printing simple but user-friendly representations of `Tuple`s, `%r` is recommended.
+             * The behavior of `%s` has not been changed to the one of `%t` or `%r` for compatibility reasons.
+             * `%s` is also kept as it is useful in debugging and unit tests.
              * ---
              *  Tuple!(int, double)[3] tupList = [ tuple(1, 1.0), tuple(2, 4.0), tuple(3, 9.0) ];
```
Comment (wilzbach) [inline std/typecons.d]: Not really related to your PR, but tests shouldn't be in plaintext. As you are already editing this file, please remove all the tests from  the ddoc header and annotate the unittests with `///` - this guarantees that the tests don't get accidentally out of date and avoids the redundancy.

**Comment by wilzbach** [comment]
I like the functionality :)  - what do other people say about using  custom format specifiers?

File: std/typecons.d (hunk 1 above)
Comment (Bolpat) [inline std/typecons.d]: No. `unittest`s in templates create a new test for each instance. Therefore we have DIP82 preventing it. As `static unittest` comes, we will do that (one day).
Note that Tuple is instantiated very often.

Comment (wilzbach) [inline std/typecons.d]: > No. unittests in templates create a new test for each instance. 

Sorry - didn't look that close. 
Yep would be great to have at some point..

**Comment by Bolpat** [comment]
Problem with documentation: see [toString](http://dtest.thecybershadow.net/artifact/website-f8ef3485d0c7d7dda9a0bc9ac55bcebca0f57e81-f1e36df795754eb97ce12f6ca8b943f0/web/phobos-prerelease/std_typecons.html#.Tuple.toString.toString.toString)
DDoc makes
> Formats Tuple with either `%s`, `%t`, `%r`, `%()inner%` or `%()inner%|sep%`.

from DDoc comment

> ``Formats Tuple with either `%s`, `%t`, `%r`, `%(`inner`%)` or `%(`inner`%|`sep`%)`.``,

but should be (italics not included, but desirable)

> Formats Tuple with either `%s`, `%t`, `%r`, `%(`_inner_`%)` or `%(`_inner_`%|`_sep_`%)`.

Can maybe someone tell me if this is a DDoc bug or intentional behavior -- or even how the result here (BTW GitHub does the job perfectly including italics) can be archived? I haven't made a big deal for DDoc for my own purposes (description is easy), but for Phobos I'd like more eye candy.

**Comment by wilzbach** [comment]
> Can maybe someone tell me if this is a DDoc bug or intentional behavior 

Looks like a bug to me, but I am not an ddoc either :/

**Comment by Bolpat** [comment]
I'm closing this and maybe open a new PR on this.

**Pull request closed without merging**